File: iosApp/iosApp/Network/NetworkManager.swift
import AuthenticationServices
import Foundation
import OSLog
import Shared

class NetworkManager {
  private var authToken: String = ""
  private let kmpClient = NetworkClient()
  private var preferences = PListManager.load()

  public func ensureAuthenticated() async throws {
    if authToken.isEmpty {
      let requestBody = TokenRequest(
        login: preferences.bankIDAuthKey, password: preferences.bankIDAuthSecret)
      let token = try await kmpClient.requestAuthToken(request: requestBody)
      self.authToken = token
    }
  }

  func getBankIdLoginURL(ngrokUrl: String) async throws -> String {
    // ALWAYS ensure authentication happens before this call
    try await ensureAuthenticated()

    let response = try await kmpClient.getBankIdAuthRedirect(
      uriRet: ngrokUrl,
      token: self.authToken
    )
    return response.redirectUrl
  }
}

File: composeApp/src/androidMain/kotlin/cz/project/ewallet/Network/NetworkManager.kt
package cz.project.ewallet

import android.util.Log
import cz.project.ewallet.model.NetworkClient
import cz.project.ewallet.BuildConfig

class NetworkManager {
        private var authToken: String = ""
        private val kmpClient = NetworkClient()
        private val TAG = "NetworkManager"

        suspend fun ensureAuthenticated() {
                if (authToken.isEmpty()) {
                        Log.d(TAG, "Requesting new token via KMP NetworkClient...")

                        val requestBody =
                                TokenRequest(
                                        login = BuildConfig.BANKID_AUTH_KEY,
                                        password = BuildConfig.BANKID_AUTH_SECRET
                                )

                        this.authToken = kmpClient.requestAuthToken(requestBody)
                        Log.d(TAG, "Successfully authenticated.")
                }
        }

        suspend fun getBankIdLoginURL(ngrokUrl: String): String {
                Log.d(TAG, "Getting BankID redirect URL...")

                ensureAuthenticated()

                val response =
                        kmpClient.getBankIdAuthRedirect(uriRet = ngrokUrl, token = this.authToken)

                return response.redirectUrl
        }
}
